### PR TITLE
Add method to extract last base AMI ID used from a particular EDP.

### PR DESCRIPTION
**Pull request by doctoryes**
https://openedx.atlassian.net/browse/TE-1814

Phase 1 of 2: @macdiesel could I get a preliminary review here? It's not finished. I'll make a couple of other notes.

### CODE REVIEW

File: tubular/tests/test_ec2.py [hunk 1]
```diff
@@ -73,6 +78,65 @@ def test_restrict_ami_to_stage(self):
...
+        }
+        fake_asg = create_asg_with_tags(
+            fake_asg_name,
+            fake_asg_tags,
+            ami_id=fake_ami_id,
+            elbs=[fake_elb]
+        )
+        self.assertEqual(ec2.ami_for_edp('foo', 'bar', 'baz'), fake_ami_id)
+
+    @unittest.skip("Test always fails due to not successfuly creating two different AMI IDs in single ELB.")
```
Comment (doctoryes) [inline tubular/tests/test_ec2.py]: I couldn't get this test to work. It was unclear whether it was moto or boto that thwarted me. Is it important enough to test this edge case?

File: tubular/tests/test_utils.py [hunk 2]
```diff
@@ -55,6 +62,17 @@ def create_asg_with_tags(asg_name, tags, ami_id="ami-abcd1234", elbs=None):
...
+
+    # Each ASG tag that has 'propagate_at_launch' set to True is *supposed* to be set on the instances.
+    # However, it seems that boto (or moto) does not properly set the tags on the instances created by the ASG.
+    # So set the tags on the ASG instances manually instead.
+    ec2_conn = boto.connect_ec2()
+    for asg in conn.get_all_groups():
+        if asg.name == asg_name:
+            asg_instance_ids = [instance.instance_id for instance in asg.instances]
+            for instance_id in asg_instance_ids:
+                ec2_conn.create_tags(instance_id, tags)
```
Comment (doctoryes) [inline tubular/tests/test_utils.py:74]: Despite adding `propagate_at_launch_time` to the tags above, I found this extra step to be necessary to get the tags on the ASG instances. Again, it was unclear if it was moto causing this issue. Thoughts?

**Comment by doctoryes** [comment]
The part that isn't done is getting the rest of the code to use this new method for base AMI selection, BTW.

File: tubular/ec2.py [hunk 3]
```diff
@@ -65,6 +67,62 @@ def get_all_load_balancers(names=None):
...
+    """
+    instance_elbs = []
+    for elb in elbs:
+        elb_instance_ids = [inst.id for inst in elb.instances]
+        if instance_id in elb_instance_ids:
+            instance_elbs.append(elb)
+    return instance_elbs
+
+
+def ami_for_edp(env, dep, play):
```
Comment (macdiesel) [inline tubular/ec2.py]: Could we name this active_ami_for_edp ?

File: tubular/tests/test_ec2.py (hunk 1 above)
Comment (macdiesel) [reply tubular/tests/test_ec2.py]: I believe your issue may be related to this: https://github.com/spulec/moto/issues/718

Since the code here: https://github.com/edx/tubular/pull/81/files#diff-62bdf1ee91eb21bc467805ff1a25b581R107 
is using multiple filters.

Comment (macdiesel) [reply tubular/tests/test_ec2.py]: Please add a note, leave the test in and skip it for now.

File: tubular/tests/test_utils.py (hunk 2 above)
Comment (macdiesel) [reply tubular/tests/test_utils.py:74]: Tagging on ASGs in MOTO is new. Are you sure your dev environment has the latest version of moto installed? 

if so I would add a note about why this is here and leave it in.

File: tubular/ec2.py [hunk 4]
```diff
@@ -65,6 +67,62 @@ def get_all_load_balancers(names=None):
...
 
+def _instance_elbs(instance_id, elbs):
+    """
+    Given an EC2 instance and ELBs, return the ELB(s) in which it is active.
+
+    Arguments:
+        instance_id(boto.ec2.instance.Reservation): Instance used to find out which ELB it is active in.
+        elbs(list(boto.ec2.elb.loadbalancer.LoadBalancer)): List of ELBs to us in checking.
+    Returns:
+        boto.ec2.elb.loadbalancer.LoadBalancer: One or more ELBs used by the passed-in instance -or- None.
```
Comment (macdiesel) [inline tubular/ec2.py]: fwiw I think we may have been doing these google doc strings a big wrong and the returns type should have backticks:
```
Args:
            param1 (str): Description of `param1`.
            param2 (:obj:`int`, optional): Description of `param2`. Multiple
                lines are supported.
            param3 (:obj:`list` of :obj:`str`): Description of `param3`.
```

**Review by macdiesel — COMMENTED** [review]
I think the testing difficulties you are having are due to moto.

Also a few suggestions.

File: tubular/ec2.py (hunk 4 above)
Comment (doctoryes) [reply tubular/ec2.py]: Will do.

File: tubular/ec2.py (hunk 3 above)
Comment (doctoryes) [reply tubular/ec2.py]: Will do.

File: tubular/tests/test_utils.py (hunk 2 above)
Comment (doctoryes) [reply tubular/tests/test_utils.py:74]: I updated it to have the most recent moto version - 0.4.30. Then I removed this section and tried tests. The test failed without this section. I'll leave it in with a note.

File: tubular/tests/test_ec2.py (hunk 1 above)
Comment (doctoryes) [reply tubular/tests/test_ec2.py]: I tried the newest version of moto (0.4.30) - no dice. Keeping the skip.

File: tubular/ec2.py (hunk 4 above)
Comment (macdiesel) [reply tubular/ec2.py]: I was incorrect about this, please ignore me.

**Review by feanil — APPROVED** [review]

**Comment by doctoryes** [comment]
@edx/pipeline-team I've added some commits since your initial reviews. Please review again.

File: scripts/retrieve_base_ami.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+from tubular import ec2  # pylint: disable=wrong-import-position
+
+logging.basicConfig(level=logging.INFO)
+
+
+@click.command()
+@click.option(
+    '--environment', '-e',
+    envvar='AMI_ENVIRONMENT',
```
Comment (macdiesel) [inline scripts/retrieve_base_ami.py]: I thought we were going to stop using environment variables unless they contained private data?

File: scripts/retrieve_base_ami.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    '--out_file',
+    help='Output file for the AMI information yaml.',
+    default=None
+)
+def retrieve_base_ami(environment, deployment, play, out_file):
+    """
+    Method used to retrieve the last base AMI ID used for an environment/deployment/play.
+    """
+    kwargs = {
+        'aws_access_key_id': os.environ.get('AWS_ACCESS_KEY_ID'),
```
Comment (macdiesel) [inline scripts/retrieve_base_ami.py]: Can we move these items in to click and pass them in via method signature?

File: tubular/ec2.py [hunk 7]
```diff
@@ -44,16 +45,20 @@ def get_all_autoscale_groups(names=None):
...
+        names (list): A list of ELB names as strings
+
     Returns:
-        a list of  of :class:`boto.ec2.elb.loadbalancer.LoadBalancer`
+        a list of :class:`boto.ec2.elb.loadbalancer.LoadBalancer`
     """
-    elb_conn = boto.connect_elb()
+    elb_conn = boto.connect_elb(
+        aws_access_key_id=kwargs.get('aws_access_key_id'),
+        aws_secret_access_key=kwargs.get('aws_secret_access_key')
```
Comment (macdiesel) [inline tubular/ec2.py]: I think boto will put these out of the environment directly:
http://boto.cloudhackers.com/en/latest/boto_config_tut.html#introduction

>  Some options, such as credentials, can also be read from environment variables (e.g. AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, AWS_SECURITY_TOKEN and AWS_PROFILE)

File: scripts/retrieve_base_ami.py (hunk 6 above)
Comment (macdiesel) [reply scripts/retrieve_base_ami.py]: I take this back, see my other comment about boto and reading these variables directly out of the environment.

>  Some options, such as credentials, can also be read from environment variables (e.g. AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, AWS_SECURITY_TOKEN and AWS_PROFILE)

File: tubular/ec2.py [hunk 8]
```diff
@@ -65,6 +70,75 @@ def get_all_load_balancers(names=None):
...
+    Returns:
+        str: Base AMI id of current active deployment for the EDP.
+    Raises:
+        MultipleImagesFoundException: If multiple AMI IDs are found within the EDP's ELB.
+        ImageNotFoundException: If no AMI IDs are found for the EDP.
+    """
+    LOG.info("Looking up AMI for {}-{}-{}...".format(env, dep, play))
+    ec2_conn = boto.connect_ec2(
+        aws_access_key_id=kwargs.get('aws_access_key_id'),
+        aws_secret_access_key=kwargs.get('aws_secret_access_key')
```
Comment (macdiesel) [inline tubular/ec2.py]: Again these should come out of the environment.

**Review by macdiesel — COMMENTED** [review]
a few comments.

File: scripts/retrieve_base_ami.py (hunk 5 above)
Comment (doctoryes) [reply scripts/retrieve_base_ami.py]: This task has dragged on for so long that we probably made the decision after I wrote this code. :) Will do.

File: scripts/retrieve_base_ami.py (hunk 6 above)
Comment (doctoryes) [reply scripts/retrieve_base_ami.py]: In my test pipeline system, the boto code was *not* using the envvars - so I added them explicitly. I'll do some more digging to figure out why.

Comment (doctoryes) [reply scripts/retrieve_base_ami.py]: I had conflated two separate issues here - all is well! I've removed the explicit reference to the AWS creds.

**Comment by doctoryes** [comment]
@macdiesel All issues addressed - ready for another round.

**Review by macdiesel — APPROVED** [review]
